File: mlp.py
import numpy as np
import pandas
import subprocess

def f(net):
    return (1 / (1 + np.exp(-net)))


def df_dnet(f_net):
    return (f_net * (1 - f_net))


def architecture(input_length=2,
                 hidden_length=2,
                 output_length=1,
                 activation=f,
                 d_activation=df_dnet):

    model = dict()
    model['input_length'] = input_length
    model['hidden_length'] = hidden_length
    model['output_length'] = output_length

    model['hidden'] = np.random.uniform(
        low=-0.5,
        high=0.5,
        size=(hidden_length,
              input_length + 1))

    model['output'] = np.random.uniform(
        low=-0.5,
        high=0.5,
        size=(output_length,
              hidden_length + 1))

    model['f'] = activation
    model['df_dnet'] = d_activation

    return model

# XOR
# 0 0 0
# 0 1 1
# 1 0 1
# 1 1 0

# Xp = [0, 1]



def backpropagation(model,
                    dataset,
                    eta=0.5,
                    threshold=1e-4):

    squaredError = 2 * threshold
    counter = 0

    while(squaredError > threshold):
        squaredError = 0

        for row in dataset:
            Xp = row[:model['input_length']]
            Yp = row[-2]
            results = forward(model, Xp)
            Op = results['f_net_o_p']

            # Calculando erro
            error = Yp - Op

            squaredError = squaredError + sum(error**2)

            delta_o_p = error * model['df_dnet'](results['f_net_o_p'])

            w_o_k_j = np.squeeze(model['output'])[:model['hidden_length']]
            delta_h_p = model['df_dnet'](
                np.asarray(results['f_net_h_p']).reshape(-1)) * \
                delta_o_p[0] * w_o_k_j

            model['output'] = model['output'] + \
                eta * np.dot(delta_o_p[0], np.append(np.asarray(
                    results['f_net_h_p']).reshape(-1), 1))
            model['hidden'] = model['hidden'] + eta * np.transpose(
                np.asmatrix(delta_h_p)) * np.append(Xp, 1)

        squaredError = squaredError / len(dataset)

        print(squaredError)

        counter += 1

    ret = dict()
    ret['model'] = model
    ret['counter'] = counter

    return ret


def forward(hidden, output, Xp):
    # Hidden layer
    net_h_p = np.dot(
        np.asmatrix(hidden),
        np.append(np.asarray(Xp), 1))
    f_net_h_p = f(net_h_p)

    # Output layer
    net_o_p = np.dot(
        output,
        np.append(np.asarray(f_net_h_p), 1))
    f_net_o_p = f(net_o_p)

    result = dict()
    result['net_h_p'] = net_h_p
    result['net_o_p'] = net_o_p
    result['f_net_h_p'] = f_net_h_p
    result['f_net_o_p'] = f_net_o_p

    return result


if __name__ == '__main__':
    '''dataset = pandas.read_csv("__data.csv")
    dataset = dataset[['std', 'minfun', 'meanfun',
                       'skew', 'maxfreq', 'iq1', 'iq3',
                       'median', 'centroid', 'minfreq',
                       'meanfreq', 'peak',
                       'kurtosis', 'maxfun', 'label']]

    dataset = dataset.values
    model = architecture(input_length=14, output_length=1, hidden_length=15)
    trained = backpropagation(model=model, dataset=dataset)
    print(dataset[0, 0:15])
    print(forward(model=trained['model'], Xp=dataset[0, 0:14]))'''
    dataset = pandas.read_csv("__data.csv")
    dataset = dataset[['std', 'minfun', 'meanfun',
                       'skew', 'maxfreq', 'iq1', 'iq3',
                       'median', 'centroid', 'minfreq',
                       'duration', 'meanfreq', 'peak',
                       'kurtosis', 'maxfun', 'label']]

    dataset = dataset.values
    hidden = pandas.read_csv("hidden.csv")
    output = pandas.read_csv("output.csv")
    import pdb; pdb.set_trace()
    result = forward(hidden=hidden, output=output, Xp=dataset[0, 0:15])
    pass

File: mlp.r
f <- function(net) {
    return ( 1/(1+exp(-net)) )
}

df_dnet <- function(f_net) {
    return (f_net * (1 - f_net))
}

mlp.architecture <- function(input.length=2,
                             hidden.length=2,
                             output.length=1,
                             activation.function=f,
                             d_activation.function=df_dnet) {

    model = list()
    model$input.length = input.length
    model$hidden.length = hidden.length
    model$output.length = output.length

    model$hidden = matrix(runif(min=-0.5, max=0.5, 
            hidden.length*(input.length+1)), 
                   nrow=hidden.length, ncol=input.length+1)

    model$output = matrix(runif(min=-0.5, max=0.5, 
            output.length*(hidden.length+1)), 
                   nrow=output.length, ncol=hidden.length+1)

    model$f = activation.function
    model$df_dnet = d_activation.function

    return (model)
}

mlp.forward <- function(model, Xp) {
    # Hidden layer
    net_h_p = model$hidden %*% c(Xp, 1)
    f_net_h_p = model$f(net_h_p)
   
   # Output layer
   net_o_p = model$output %*% c(as.numeric(f_net_h_p), 1)
   f_net_o_p = model$f(net_o_p)

   # Results
   ret = list()
   ret$net_h_p = net_h_p
   ret$net_o_p = net_o_p
   ret$f_net_h_p = f_net_h_p
   ret$f_net_o_p = f_net_o_p

   return (ret)
}

mlp.backpropagation <- function(model,
                                dataset,
                                eta=0.1,
                                threshold=1e-4) {

    squaredError = 2 * threshold
    counter = 0

    while(squaredError > threshold) {
        squaredError = 0
 
        for (p in 1:nrow(dataset)) {
            Xp = as.numeric(dataset[p, 1:model$input.length])
            Yp = as.numeric(dataset[p,
                    (model$input.length+1):ncol(dataset)])

            results = mlp.forward(model, Xp)
            Op = results$f_net_o_p

            # Calculando o erro
            error = Yp - Op

            squaredError = squaredError + sum(error^2)

            delta_o_p = error * model$df_dnet(results$f_net_o_p)
        
            w_o_kj = model$output[,1:model$hidden.length]
            delta_h_p = 
                as.numeric(model$df_dnet(results$f_net_h_p)) *
                    (as.numeric(delta_o_p) %*% w_o_kj)

            model$output = model$output +
                eta*(delta_o_p%*%as.vector(c(results$f_net_h_p,1)))
            model$hidden = model$hidden +
                eta*(t(delta_h_p) %*% as.vector(c(Xp,1)))
        }
        squaredError = squaredError / nrow(dataset)

        cat("Erro médio quadrado = ", squaredError, "\n")

        counter = counter + 1
    }
    cat("Base treinada Daniel, SEU GOSTOSO!")
    ret = list()
    ret$model = model
    ret$counter = counter

    return (ret)
}


dataset = read.csv('__data.csv', header=F, skip=1)
dataset <- dataset[c(1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13, 14, 15, 16, 8)]
ids = sample(1:nrow(dataset), size=50)
training = dataset[ids,]
test = dataset[-ids,]
model = mlp.architecture(input.length=15, output.length=1, hidden.length=16)
trained = mlp.backpropagation(dataset=training, model=model, eta=0.5)
write.csv(trained$model$hidden, file="hidden.csv", row.names=FALSE)
write.csv(trained$model$output, file="output.csv", row.names=FALSE)
